Fix various bugs in the Bloom Player implementation

rtibbles opened the pull request:
## Summary
* Does small clean up of unneeded conditional logic in the bloom player vue component to allow bloomd files to be rendered
* Fixes previously unnoticed bug where URLs with spaces in them, or URI encoded URLs would not get properly replaced by kolibri-zip - adds regression tests to ensure behaviour
* Migrates all the bloom player specific code that handles mapping audio files into the bloom player code, use the extensibility of the kolibri-zip module to allow this.
* Move the HTML style attribute handling into the main kolibri-zip DOMMapper
* Also fix URL spaces in names and URI encoding issues for DOM parsing too
* Update the DOMMapper to handle URLs in inline `<style>` blocks
* Do a small refactor to make code more readable, and give slight performance improvement on small DOMs
* Use URLSearchParameters to handle passing options to the bloom player iframe
* Add more parameters to hide the internal toolbar (defer to our own toolbar)
* Update bloom player to the latest version from upstream
* Include a small styling patch to make the previous/next buttons smaller
* Add to hashi README to describe how to update the bloom-player

## References
Fixes #12728 
Fixes #12726 
Fixes #12725 
Fixes #12727 

## Reviewer guidance
Verify that bloom reader is:
* still functioning as expected
* bloomd files are now renderable
* the internal bloom player controls are hidden
* all images are properly displayed inside books (when compared with epubs)
* the previous/next buttons are now a more reasonable size when hovered.

----

## Testing checklist

- [ ] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Critical and brittle code paths are covered by unit tests


## PR process

- [ ] PR has the correct target branch and milestone
- [ ] PR has 'needs review' or 'work-in-progress' label
- [ ] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

## Reviewer checklist

- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

rtibbles committed:
Small cleanup to ensure bloomd files render properly.

rtibbles committed:
Handle spaces and URI encoded urls in CSS.

rtibbles committed:
Cleanup bloom specific mapper code.
Use custom mappers.
Fix url edge cases in unzipping.

rtibbles committed:
Use URLSearchParams to form URL query parameters.

rtibbles committed:
Update bloom player.
Apply our patch to reduce forward backward button size.

rtibbles committed:
Add instructions for updating bloom reader files.

rtibbles committed:
Handle file URL replacement in inline style blocks in HTML.

rtibbles committed:
Add tests and handling for a broader range of encoded characters.

rtibbles committed:
Add tests and handling for srcset attribute.

rtibbles committed:
Refactor for small performance improvement on small pages, and improve code readability.

pcenov commented:
Hi @rtibbles, I confirm that .bloomd files are now renderable and I also confirm the rest of the bullet points in the reviewer guidance. Issues observed during this round of QA which can be reported separately:

1. Some files (both of .bloomd and .bloompub type) cannot be rendered in the player and I am seeing the following error in the console: 

```
BloomRunner-859fb260cac123ba42b8.bundle.js:1 Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'split')
    at BloomRunner-859fb260cac123ba42b8.bundle.js:1:4138
    at e.value (BloomRunner-859fb260cac123ba42b8.bundle.js:1:4247)
    at 627-8ce4f4671fde42067b92.bundle.js:1:11867
    at async Promise.all (learn/index 0)
```

Examples are` 篾匠` and `157 God Gives Solomon Wisdom` which can be imported from the following channel: `tonih-nabul`

![error](https://github.com/user-attachments/assets/5de6e0bd-70b1-488d-898e-a5cdc7bb356a)

2. All bloompub files always start from the beginning, so the progress made by the learner is not being saved.
3. The resource completed modal is never shown but I do see the 'Completed' icon

https://github.com/user-attachments/assets/3b97fd2a-fec7-4311-b509-5ac8989efbd4

4. bloompub files with audio are shown with a huge play button so that the audio can be started but there's no option to stop it. See `WEB Ivile nɛ Ɛcʉlʉ` for that:

https://github.com/user-attachments/assets/04038d83-14de-4c17-af69-0b7d560d0255

5. Both arrows look as if they are disabled:

https://github.com/user-attachments/assets/d83df91e-c02d-462b-a1e9-a48d772e9eeb

rtibbles commented:
I think I'll fix 1 and 3 in the course of this PR, and we can file the others as follow up. Thanks for the channel reference!

rtibbles committed:
Separate out DOM queries to avoid having to try to duck type different attribute matches.

rtibbles committed:
Reinstate internal app bar to allow pausing of books.

rtibbles commented:
Oh, I am addressing 4, as that was actually caused in this PR, so I just need to slightly undo something I did earlier.

rtibbles committed:
Fix completion behaviour for bloom player.
Flyby fix to ensure that hashi progress also emits finished event in HTML5 Player.

rtibbles committed:
Add more edge case handling for path replacement with tests.

rtibbles commented:
@pcenov 1, 3, and 4 should now be addressed.

2 is not implemented in the bloom player, so would require a fair amount of work to get that working. 5 is more tractable, but would need a small update in the player code.

pcenov commented:
Hi @rtibbles I confirm that points 1, 3, and 4 are addressed now and there are no new issues observed.
I have also filed follow-up issues for 2 and 5.

rtibbles commented:
Thank you, @pcenov!

marcellamaki commented on kolibri/core/content/static/bloom/bloomplayer.htm:
```diff
@@ -1,13 +1,16 @@
+        <meta charset="UTF-8" />
+    </head>
+    <body style="background-color: #2e2e2e;">
+        <div id="root"><span style="color: #d65649">Loading Bloom Player...</span></div>
```
i know this is pre-existing, but do we want to translate this? could we translate it? (I realize it's not a vue file and therefore that might be more complicated. i haven't really thought through fully how we would do the translation)

radinamatic replied:
I believe we should be translating this, yes 👍🏽 

We did some string extraction from HTML, but not in Kolibri repo, one of the installers iirc...
Do our i18n scripts here support extraction from HTML? 🤔

rtibbles replied:
We can't translate it, without a bunch of additional tooling. We'd be better off just deleting it.

marcellamaki commented on kolibri/plugins/bloompub_viewer/assets/src/views/BloomPubRendererIndex.vue:
```diff
@@ -162,58 +156,21 @@
         this.$emit('error', err);
       });
-      let storageUrl = this.defaultFile.storage_url;
-      if (!this.isBloom) {
```
all of this code deletion is because in the bloompub renderer index, we wouldn't be rendering non-bloom files? or is it something else

rtibbles replied:
Just that, yes.

marcellamaki commented on kolibri/plugins/bloompub_viewer/assets/src/views/BloomPubRendererIndex.vue:
```diff
@@ -162,58 +156,21 @@
       this.$emit('stopTracking');
     },
-    methods: {
-      recordProgress() {
```
and removing this is because the progress tracking isn't working, and we're fixing that in follow up?

rtibbles replied:
Because this method is only used for periodic polling of progress recording - whereas all progress tracking is driven by events from the bloom player for this (this was copy pasted from the HTML5 renderer).

marcellamaki commented on packages/hashi/README.md:
```diff
@@ -173,3 +173,11 @@ H5P Static Files
+
+This code is currently generated from https://github.com/learningequality/bloom-player (specifically the 'patched' default branch).
+
+To regenerate, the repository should be cloned, and `yarn run build` run within the context of that repository to regenerate the new assets. All the files put into `dist` should then be copied into `kolibri/core/content/static/bloom` in the Kolibri repository. Any previously existing hash named files can be deleted and replaced by the new hash named files.
```
this might be somewhat out of scope for the readme, but what are the scenarios when we would want to regenerate the static files?

rtibbles replied:
When there is an update from the bloom repo. Once I have finished the auto-updating H5P action and got that working, I am going to make one for bloom too.

marcellamaki commented on packages/hashi/src/Bloom/BloomRunner.js:
```diff
@@ -66,67 +132,37 @@ export default class BloomRunner {
     }
   }
 
-  processContent() {
```
is this

> Migrates all the bloom player specific code that handles mapping audio files into the bloom player code, use the extensibility of the kolibri-zip module to allow this

?

rtibbles replied:
Yes - the `processContent` method was doing post-hoc processing because the mapping logic wasn't being setup for the zip file in the first place. All of what was happening here is now happening inside the BloomDOMMapper class that is being defined in here, in order to make it happen while we are extracting files from the zip file.

marcellamaki reviewed (commented):
a few questions

rtibbles committed:
Update bloom player to latest, remove loading state from HTML.

rtibbles commented on packages/kolibri-zip/src/fileUtils.js:
```diff
@@ -68,117 +81,109 @@ const domParser = new DOMParser();
-    ),
+  // Get paths from URL attributes (src, href)
+  const urlPaths = flatten(
+    urlAttributes.flatMap(attr => {
```
This is an error, this should be `map` not `flatMap` as `flatMap` is only supported from iOS Safari 12 and onwards: https://caniuse.com/?search=flatMap

marcellamaki reviewed (approved):
I think this is good to go @rtibbles - thank you! comprehension questions/comments all addressed, and no blocking concerns
